### PR TITLE
show the last entity shown in preview

### LINKED ISSUE #648: Form Designer preview
- [ ] After closing preview, in form designer show the last entity shown in preview

*Check if it makes sense to have the nodeDefUuidPage separated from data entry*

### PR DESCRIPTION
**Pull request by ramirobg94**
## Related Issue

Resolves #648 

## Brief description of the changes proposed an/or how the issue(s) has(have) been solved.

- remove the action handlers that modify the current page in the survey form to show the last entity shown in preview
- Add a new action to be triggered when the form is unmounted to reset the current page

## How has this been tested

Manual testing

##### Do you consider this PR needs further testing?
- [ ] No
- [x] Yes

## Types of changes

- [ ] Docs change 
- [ ] Refactoring 
- [ ] Dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Disclaimer

https://user-images.githubusercontent.com/11272068/108964111-5855b900-767b-11eb-87f9-f8e707f246a1.mov

### CODE REVIEW

File: webapp/store/ui/surveyForm/actions.js [hunk 1]
```diff
@@ -20,6 +20,9 @@ export const formReset = 'survey/form/reset'
 
 export const resetForm = () => (dispatch) => dispatch({ type: formReset })
 
+export const formClean = 'survey/form/clean'
+export const cleanForm = () => (dispatch) => dispatch({ type: formClean })
```
Comment (minotogna) [inline webapp/store/ui/surveyForm/actions.js:24]: clean might be confusing. cant we use reset already implemented ?

**Review by minotogna — CHANGES_REQUESTED** [review]

File: webapp/components/survey/SurveyForm/SurveyForm.js [hunk 2]
```diff
@@ -75,6 +75,14 @@ const SurveyForm = (props) => {
     }
   }, [])
 
+  useEffect(() => {
+    return () => {
+      if (entry && !preview) {
+        dispatch(SurveyFormActions.resetForm())
```
Comment (minotogna) [inline webapp/components/survey/SurveyForm/SurveyForm.js:81]: do we need to reset the form in both surveyForm and record?

Comment (ramirobg94) [reply webapp/components/survey/SurveyForm/SurveyForm.js:81]: States:
A - Form Designer: `{entry: false, preview: false, edit: true}`
B - Form Designer-preview: `{entry: true, preview: true edit: false}`
C- Entry record: `{entry: true, preview: false edit: false}`

Goal: (After closing preview, in form designer show the last entity shown in preview)

Cases: 
A - B -> reset ( not reset if we like to move from designer to preview and keep the page)
B - A -> not reset -> fix
A - C -> reset
B - C -> reset
C - X -> reset

Comment (minotogna) [reply webapp/components/survey/SurveyForm/SurveyForm.js]: B->C and C->B do not exists .

cant you reset the form only when you unmount the surveyForm?

Comment (ramirobg94) [reply webapp/components/survey/SurveyForm/SurveyForm.js]: B-C is possible, go to Designer, open Preview go-to records
C-B agree

If we only reset on unmount of SurveyForm is going to reset always: SurveyFrom is used on the three states.

File: webapp/components/survey/SurveyForm/SurveyForm.js [hunk 3]
```diff
@@ -75,6 +77,13 @@ const SurveyForm = (props) => {
     }
   }, [])
 
+  useHistoryListen((location) => {
+    // user enters in records the form is reset
+    if (matchPath(location.pathname, { path: appModuleUri(dataModules.records) })) {
+      dispatch(SurveyFormActions.resetForm())
+    }
+  }, [])
+
```
Comment (ramirobg94) [inline webapp/components/survey/SurveyForm/SurveyForm.js:86]: Move to form Designer

File: webapp/components/survey/Record/store/useLocalState.js [hunk 4]
```diff
@@ -59,7 +59,9 @@ export const useLocalState = () => {
     }
 
     // Reset form
-    dispatch(SurveyFormActions.resetForm())
+    if (!preview) {
+      dispatch(SurveyFormActions.resetForm())
+    }
```
Comment (ramirobg94) [inline webapp/components/survey/Record/store/useLocalState.js:64]: check if we are able to move to FormDesigner Unmount

File: webapp/components/survey/SurveyForm/SurveyForm.js [hunk 5]
```diff
@@ -75,6 +77,13 @@ const SurveyForm = (props) => {
     }
   }, [])
 
+  useHistoryListen((location) => {
+    // user enters in records the form is reset
+    if (matchPath(location.pathname, { path: appModuleUri(dataModules.records) })) {
+      dispatch(SurveyFormActions.resetForm())
```
Comment (minotogna) [inline webapp/components/survey/SurveyForm/SurveyForm.js:83]: can you check if dispatching this action only on the FormDesigner unmount might work?

File: webapp/components/survey/Record/store/useLocalState.js [hunk 6]
```diff
@@ -59,7 +59,9 @@ export const useLocalState = () => {
     }
 
     // Reset form
-    dispatch(SurveyFormActions.resetForm())
+    if (!preview) {
+      dispatch(SurveyFormActions.resetForm())
```
Comment (minotogna) [inline webapp/components/survey/Record/store/useLocalState.js:63]: can you check if this is needed? provably not ?

File: webapp/components/survey/Record/store/useLocalState.js (hunk 4 above)
Comment (ramirobg94) [reply webapp/components/survey/Record/store/useLocalState.js:64]: check if we can remove this.

File: webapp/components/survey/Record/store/useLocalState.js (hunk 6 above)
Comment (ramirobg94) [reply webapp/components/survey/Record/store/useLocalState.js]: cleaned :)

**Review by minotogna — DISMISSED** [review]

**Review by minotogna — DISMISSED** [review]

**Comment by ramirobg94** [comment]
I applied some changes because navigation between surveyForm to nodeDef wasn't handled

**Review by SteRiccio — APPROVED** [review]

**Pull request merged**